keps/sig-cluster-lifecycle/image-builder/20190612-image-builder.md

```diff
@@ -0,0 +1,122 @@
+* To provide upgrade or downgrade semantics.
+* To provide guarantees that the software installed provides a fully functional system.
+* To prescribe the hardware architecture of the build system.
+* To create images from scratch.  
+    * The purpose of the tool is to take an existing image and make it Kubernetes ready.
+
+## Proposal 
+The Image Builder will start from one image in a supported format and create a new image in the same format specifically for the purpose of creating Kubernetes clusters.  In surveying the landscape of tooling it becomes readily apparent that there are a plethora of tools that provide an opinionated end-to-end user story around image creation, but we’ve observed it can be decomposed into a series of steps, or phases.  By decomposing the problem we can provide a rallying point for different tools to integrate, and provide the kubernetes ecosystem with a common utility and user experience across those tools.
+
+As a precondition the Image Builder will require a bootable disk image as an input, with native support for the cloud images published by the supported distributions.  However any external process or tool can be used to create the initial disk image from other sources including [ISO](https://packer.io)’s, file trees and [docker](https://github.com/iximiuz/docker-to-linux) images. Existing disk images can also be customized using tools like [virt-customize](http://libguestfs.org/virt-customize.1.html) before being fed into the Image Builder.
+
+**NOTE:** It should be noted that this document is intentionally high level and purposefully omits design choices which should be made at a later date once the subproject is further along in its lifecycle. 
+
+### Phases 
+#### Phase 0 (Base Image)
+
+Lay down the initial base image.  Often times this can be some form of certified base image from a vendor or IT team.   **NOTE:** It is not a goal of this project to take on creation of those initial images.
+
+**Input:**
+
+`--disk-image` - A local or remote path to a libvirt/qemu supported image that a user or provider creates. (raw/qcow/vmdk etc.)
+
+Images are verified and cached by looking for a SHA256SUMS, sha256sum.txt file in the same directory as the image
+
+* [Ubuntu](https://cloud-images.ubuntu.com/bionic/current/)
+* [Fedora](https://alt.fedoraproject.org/cloud/)
+* [Debian](https://cloud.debian.org/images/openstack/) and a [comparison](https://wiki.debian.org/Cloud/SystemsComparison) of the types
+* [CentOS](https://cloud.centos.org/centos/7/images/)
+* [Amazon Linux](https://cdn.amazonlinux.com/os-images/current/kvm/)
+
+**Output:** Running shell inside the root filesystem
+
+Phase 0 will kickoff Phase 1, by either chrooting into the disk or using nested virtualization to boot the image and then SSH into it.
```

Comment:
We might want to say "for example by chrooting into the disk..." because ideally we wouldn't need to do any execution and could instead just lay down files.